lib/spack/spack/package.py

```diff
@@ -2195,6 +2195,85 @@ def possible_dependencies(*pkg_or_spec, **kwargs):
     return visited
 
 
+#: Registers which are the detectable packages, by repo and package name
+detectable_packages = collections.defaultdict(list)
+
+
+def detectable(decorated_cls):
```

Comment:
What's the advantage of making this a decorator vs. adding these methods as default implementations to `Package` (possibly as a mixin). Packages have different features based on their functionality already and IMO as a general rule it is idea to implement them as mixins vs. decorators to avoid decorator ordering conflicts.

Reply:
> What's the advantage of making this a decorator vs. adding these methods as default implementations to Package (possibly as a mixin). 

It can obviously be done in both ways (decorator or mixin), but I think a decorator would be preferable here since it keeps build related properties (the current base classes and mixins are all related to behavior at build time somehow) separate from other types of behavior. 

> Packages have different features based on their functionality already and IMO as a general rule it is idea to implement them as mixins vs. decorators to avoid decorator ordering conflicts.

Ordering conflicts are also possible with mixins, since the MRO depends on the order in which base classes are declared . Besides that, this particular decorator is simple enough not to have ordering issues.

Reply:
> I agree with @scheibelp that this would be better as a Mixin, or better as part of the PackageBase class. [ ... ] Every package in Spack should be detectable if it defines the appropriate interface.

I think this is a point of lesser importance with respect to the transitioning of configuration files, but I  disagree with that:

1. Having default implementations in `PackageBase` would bloat that class even more
2. Using a mixin would be preferable if we had to leverage `PackageBase` metaclass, but here we don't

Another point is that while almost every package needs to be buildable from sources (current mixins deal with that) I think that in the long term only a small fraction of the packages will be detectable, mainly build tools and vendor provided libraries. If we foresee e.g. to have ~200 packages that can be detected that would be less than 5% in our current builtin repo.

Another point to consider is that a decorator makes implementing certain extensions easier. For instance an extension I'd like to work on in a following PR (after v0.16.0) are shorthands of the form:
```console
$ spack external find compilers
$ spack external find build-tools
$ spack external list compilers
==> Detectable compilers per repository
Repository: builtin
    apple-clang   cce  fj  gcc  intel  llvm  pgi  xlc  xlf
```
where "compilers" and "build-tools" are collections of detectable packages. With a decorator one can have a neat interface like:
```python
@detectable('compilers,build-tools')
class Gcc(AutotoolsPackage):
    ...
```
and construct a dictionary of categories during the definition of the class in an easy way. With mixins this would look similar to:
```python
class Gcc(AutotoolsPackage, DetectablePackage):
    detection_categories = ['compilers', 'build-tools']
```
and the implementation needs either to hook into metaclasses (to be done during class definition) or require a second pass of inspection for attributes. So, while all of this is doable in multiple ways I think there are reasons to prefer decorators.

Reply:
@scheibelp @becker33 In e42bb45 I went with a different implementation than the ones suggested here but to the user it has the same effect of having everything in `PackageBase` i.e. the decorator step or adding a Mixin is not needed anymore. That possibly combines what I see good with decorators (don't clutter with methods packages that don't want to be detectable, room for doing bookkeeping at class definition time, keep all the detection methods together and separated from unrelated code) with the simplification of the process you are requiring. 

Since it is something different from the options proposed I won't close the thread right away. Let me know if this is considered resolved or if there are more concerns.

Reply:
I personally consider this fine

@becker33?

Reply:
I haven't dove into the details, but I agree in principle that seems fine.